drivers/bluetooth/hci/userchan_bottom.c

```diff
@@ -0,0 +1,99 @@
+/**
+ * Copyright (c) 2025 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * "Bottom"/Linux/Host side of the userchannel driver. This file is built in the native_simulator
+ * runner context with the host libC, but it's functionality can be called from the "embedded" side
+ */
+
+#include <stdbool.h>
+#include <string.h>
+#include <errno.h>
+#include <poll.h>
+#include <unistd.h>
+#include <sys/socket.h>
+#include <limits.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <nsi_errno.h>
+
+#define BTPROTO_HCI      1
+#define HCI_CHANNEL_USER 1
+
+struct sockaddr_hci {
+	sa_family_t    hci_family;
+	unsigned short hci_dev;
+	unsigned short hci_channel;
+};
+
+bool user_chan_rx_ready(int fd)
+{
+	struct pollfd pollfd = {.fd = fd, .events = POLLIN};
+
+	return (poll(&pollfd, 1, 0) == 1);
+}
+
+int user_chan_is_ipaddr_ok(char ip_addr[])
+{
+	struct in_addr addr;
+
+	return inet_pton(AF_INET, ip_addr, &addr);
+}
+
+int user_chan_hci_open(unsigned short bt_dev_index)
+{
+	int fd;
+	struct sockaddr_hci addr;
+
+	fd = socket(PF_BLUETOOTH, SOCK_RAW | SOCK_CLOEXEC | SOCK_NONBLOCK, BTPROTO_HCI);
+	if (fd < 0) {
+		return -nsi_errno_to_mid(errno);
+	}
+
+	(void)memset(&addr, 0, sizeof(addr));
+	addr.hci_family = AF_BLUETOOTH;
+	addr.hci_dev = bt_dev_index;
+	addr.hci_channel = HCI_CHANNEL_USER;
+
+	if (bind(fd, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
+		int err = -nsi_errno_to_mid(errno);
+
+		close(fd);
+		return err;
+	}
+
+	return fd;
+}
+
+
+int user_chan_socket_open(char ip_addr[], unsigned int port)
```

Comment:
Thanks for doing the split. Although this now highlights even stronger why it's counterintuitive to have network sockets mixed into this driver: user channel sockets are a dedicated Bluetooth socket type in Linux, and don't really have anything in common with network sockets. Yet, this API above is called `user_chan_socket_*()`, which is pretty misleading. I realize you want to keep the namespace the same for these helpers since they're all related to the userchan HCI driver, but I wonder if this could be renamed to e.g. `user_chan_net_connect()` to highlight the difference in socket type and well as operation type (you're not just opening something rather actually establishing a connection)?

Reply:
In fact, I think `user_chan_socket_open()` is the correct name for the function that opens the real user channel socket (the one you now call `user_chan_hci_open()`)

Reply:
Done